src/Analyser/MutatingScope.php

```diff
@@ -2456,13 +2456,25 @@ public function enterTrait(ClassReflection $traitReflection): self
 		if (count($traitNameParts) > 1) {
 			$namespace = implode('\\', array_slice($traitNameParts, 0, -1));
 		}
+
+		$traitContext = $this->context->enterTrait($traitReflection);
+		$classReflection = $traitContext->getClassReflection();
+		if ($classReflection === null) {
+			throw new ShouldNotHappenException();
+		}
+		$thisHolder = ExpressionTypeHolder::createYes(new Variable('this'), new ThisType($classReflection, null, $traitReflection));
+
 		return $this->scopeFactory->create(
-			$this->context->enterTrait($traitReflection),
+			$traitContext,
 			$this->isDeclareStrictTypes(),
 			$this->getFunction(),
 			$namespace,
-			$this->expressionTypes,
-			$this->nativeExpressionTypes,
+			array_merge($this->expressionTypes, [
+				'$this' => $thisHolder,
```

Comment:
No need for array_merge. Simply do this before:

```php
$expressionTypes = $this->expressionTypes;
$expressionType['$this'] = $thisHolder;
```

And the same thing for nativeExpressionTypes.

Reply:
I see. I was inspired by another place where it was done this way. fixed both in [749fbab](https://github.com/phpstan/phpstan-src/pull/2045/commits/749fbab85d2194dcd3be0fb5d1c290d62bad7dc0)